From 0d2b359754bbfe8678b8e39cea044df6722e4395 Mon Sep 17 00:00:00 2001
From: Ricardo Neri <ricardo.neri-calderon@linux.intel.com>
Date: Thu, 19 Mar 2015 10:55:32 -0700
Subject: [PATCH 05/16] use grub_printf

---
 grub-core/loader/efi/chainloader.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/grub-core/loader/efi/chainloader.c b/grub-core/loader/efi/chainloader.c
index 892da4f..d4e2a7a 100644
--- a/grub-core/loader/efi/chainloader.c
+++ b/grub-core/loader/efi/chainloader.c
@@ -504,7 +504,7 @@ grub_cmd_chainloader (grub_command_t cmd __attribute__ ((unused)),
       goto fail;
     }
 
-  grub_error (GRUB_ERR_BAD_OS, "file size [0x%x]", size);
+  grub_printf ("file size [%d]", (int)size);
 
   status = efi_call_6 (b->load_image, 0, grub_efi_image_handle, file_path,
 			  (void *) ((grub_addr_t) address), size,
@@ -519,7 +519,7 @@ grub_cmd_chainloader (grub_command_t cmd __attribute__ ((unused)),
 	grub_efi_shim_lock_t *c;
 
         c = grub_efi_locate_protocol (&shim_protocol_guid, 0);
-        grub_error (GRUB_ERR_BAD_OS, "protocol find [0x%x]", c);
+        grub_printf ("protocol find [0x%p]", (void *)c);
       }
 
       goto fail;
-- 
1.9.1

