From f3493bf3bd2453ac785ec902765a23f78d70275d Mon Sep 17 00:00:00 2001
From: Ricardo Neri <ricardo.neri-calderon@linux.intel.com>
Date: Fri, 20 Mar 2015 14:13:09 -0700
Subject: [PATCH 11/16] define entry point

---
 grub-core/loader/efi/chainloader.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/grub-core/loader/efi/chainloader.c b/grub-core/loader/efi/chainloader.c
index cefb2cb..ac21b40 100644
--- a/grub-core/loader/efi/chainloader.c
+++ b/grub-core/loader/efi/chainloader.c
@@ -53,6 +53,7 @@ static grub_efi_char16_t *cmdline;
   }
 
 
+static grub_efi_status_t (*entry_point) (grub_efi_handle_t image_handle, grub_efi_system_table_t *system_table);
 //
 // Additional PE32+ Machine type for EFI images
 //
@@ -880,6 +881,9 @@ grub_cmd_chainloader (grub_command_t cmd __attribute__ ((unused)),
 		grub_printf("Relocation success: [%ld]\n", status);
 	}
 
+	entry_point = (void *)image_address(buffer, context.ImageSize, context.EntryPoint);
+
+	grub_printf("entry point is[0x%lx]\n", entry_point);
 	grub_printf("I made it here!!\n");
 	grub_error (GRUB_ERR_BAD_OS, "cannot load image [%d]", status);
       }
-- 
1.9.1

